### DISCUSSION TITLE
Match Media Empty Theme

### DISCUSSION THREAD

**Discussion by mod-flux** [system-ui/theme-ui#1605] [Q&A] [answered]
I'm having issues with the [match-media](https://github.com/system-ui/theme-ui/tree/develop/packages/match-media) package and was curious if anyone has come across any issues or know a fix. I've searched GitHub but it doesn't appear anyone has come across this issue before.

I'm using Theme-UI in a Next.js project and whilst the `useResponsiveValue()` and `useBreakpointIndex()` functions work in my component, I noticed that my custom breakpoints were being ignored and only default ones were being used. It's worth mentioning in all other uses of the `[]` syntax for breakpoint styling, works as expected with my custom `breakpoints` in my `theme` object.

I thought perhaps the `context.theme` didn't exist, which is what `match-media` relies on for the `theme` object. I tested this theory by doing the following:

```
import { useThemeUI } from 'theme-ui'

const Component = () => {
  const {theme} = useThemeUI()
  
  console.log(theme) // correctly logs my custom theme containing custom breakpoints

  const foo = useResponsiveValue([false, null, true]) // receives a theme object of {} and therefore uses the default breakpoints :-(
}
```

Am I missing something anyone can suggest?

**Comment by hasparus** [comment, marked as answer]
@mod-flux, could you paste your lockfile here? If a context doesn't exist it's most often a version mismatch between Theme UI or Emotion.

**Reply by mod-flux** [reply]
@hasparus I've had a quick look but nothing is really jumping out, here it is:

[yarn.lock.zip](https://github.com/system-ui/theme-ui/files/6224318/yarn.lock.zip)

Any advice you can give would be great.

**Reply by hasparus** [reply]
You have @emotion/core 10 and @emotion/react 11. Yes, this is the same package. Thus you have both v10 and v11 of @emotion/cache. Are you testing your components in Storybook, or do you experience this problem in Next?

We could try to reproduce this issue in CodeSandbox.

**Reply by mod-flux** [reply]
@hasparus the issue is present in both the storybook version of the component in question and the Next.js version. I will try to figure out why I have both of those versions installed at the same time. Will also try and reproduce in CodeSandbox!

**Reply by hasparus** [reply]
> I will try to figure out why I have both of those versions installed at the same time.

Storybook depends on Emotion 10, but it shouldn't really affect your Next project... Hmm... Have you tried console.logging `useThemeUI()` imported from `@theme-ui/core`? I'm grasping at straws.

**Reply by mod-flux** [reply]
@hasparus I'm going to do more debugging and digging and get back to you. Atleast I know I'm not missing something more obvious, gives me a place to start. :-)

**Reply by mod-flux** [reply]
The plot thickens @hasparus! Importing `useThemeUI` either from `theme-ui` or `@theme-ui/core` in storybook and logging out the returned context all I get is `{__EMOTION_VERSION__: "11.1.5", theme: {}}`. So for whatever reason, it's not available in my Storybook components either.

I compile my components using Babel and make them available to the website repo in a traditional Yarn workspaces monorepo setup so logic would suggest if the compile step in my Storybook repo is unable to get the theme then once compiled, anything using it will also be missing the theme (my site). In theory, if I can get it working in storybook, it should technically work on my website. My .babelrc.js currently looks like:

```
module.exports = {
  presets: [
    [
      'next/babel',
      {
        'preset-env': {
          useBuiltIns: 'entry',
          corejs: '3.6',
        },
        'preset-react': {
          runtime: 'automatic',
          importSource: 'theme-ui',
        },
      },
    ],
  ],
  plugins: ['@emotion', '@babel/plugin-transform-runtime'],
}
```

**Reply by hasparus** [reply]
Okay, another thing I'd try is opening React devtools and counting all context providers from Theme UI to see where's my theme.

**Reply by mod-flux** [reply]
@hasparus figured out the problem or at least a solution. We've rolled our own patch-package for `@theme-ui/media-match` to fix it. Essentially in just changing the imports at the top of the `index.js` in `match-media` from `import { useThemeUI } from '@theme-ui/core'` to `import { useThemeUI } from 'theme-ui'` it all works as intended. Apparently due to the library importing from `@theme-ui/core` it was using a different context to that of our website which was using `theme-ui`, oddly.

We have no idea why this works but appears to do the trick.

**Reply by hasparus** [reply]
This would suggest that we have multiple copies of `@theme-ui/core` in node_modules :thinking:

But `useThemeUI` is only defined once (in core) and reexported in `theme-ui`...

I'm not sure if it was the cause of the problem, but match-media had broken peer deps.

![image](https://user-images.githubusercontent.com/15332326/113061541-7fd7fe00-91b2-11eb-8ad2-77bf4c0ba62f.png)

I've updated this peerDependency, and I'm [deploying](https://github.com/system-ui/theme-ui/runs/2231895827) a new canary release, but I'm not sure if it helps.

While I still have you here, could you tell me if the following if both `useThemeUI` imports are referentially equal in your project? 

```tsx
import { useThemeUI as one } from 'theme-ui'
import { useThemeUI as two } from '@theme-ui/core'

console.log(one === two)
```

**Reply by mod-flux** [reply]
@hasparus I have exciting news! Your fix appears to have fixed the issue and furthermore, I've tested it using the above method, and everything matches in Storybook and our site. It looks like the lower version of the `peerDepedency` was throwing things out :-)

Thanks for fixing it, appreciate it!

**Reply by hasparus** [reply]
Thanks! I'm glad it works for you! Though that's bad news... After so many years it turns out I don't know how peer deps work :facepalm:
